src/PluginStartupRegistration.cpp

```diff
@@ -100,31 +208,68 @@ const std::vector<wxString>& PluginStartupRegistration::GetFailedPluginsPaths()
    return mFailedPluginsPaths;
 }
 
-void PluginStartupRegistration::Run()
+void PluginStartupRegistration::Run(std::chrono::seconds timeout)
 {
-   auto dialog = BasicUI::MakeProgress(XO("Searching for plugins"), XO(""));
-   ProcessNext();
-   while(mValidator)
+   PluginScanDialog dialog(nullptr, wxID_ANY, XO("Searching for plugins"));
+   wxTimer timeoutTimer(&dialog, OnPluginScanTimeout);
+   mScanDialog = &dialog;
+   mTimeoutTimer = &timeoutTimer;
+   mTimeout = timeout;
+
+   auto cleanup = finally([this]
    {
-      const auto message = TranslatableString { mPluginsToProcess[mCurrentPluginIndex].first, { } };
-      //Update UI
-      if(dialog->Poll(mCurrentPluginIndex, mPluginsToProcess.size(), message) != BasicUI::ProgressResult::Success)
-         Stop();
-      //AsyncPluginValidator uses event loop for internal message
-      //delivery, but ProgressDialog::Poll implementation does not call
-      //wxApp::Yield each time, which may result in too long CPU stalls
-      if(wxTheApp->HasPendingEvents())
-         wxTheApp->Yield();
+      mTimeoutTimer = nullptr;
+      mScanDialog = nullptr;
+   });
+
+   dialog.Bind(wxEVT_BUTTON, [this](wxCommandEvent& evt) {
+      evt.Skip();
+      if(evt.GetId() == wxID_IGNORE)
+         Skip();
+   });
+   dialog.Bind(wxEVT_TIMER, [this](wxTimerEvent& evt) {
+      if(evt.GetId() == OnPluginScanTimeout)
+      {
+         if(mValidator->InactiveSince() < mRequestStartTime)
+            Skip();
+         //else
+         //   wxMessageBox("Please check for plugin popups!");
+      }
       else
-         //Seems like we do have no events yet, save CPU cycles then
-         std::this_thread::sleep_for(std::chrono::milliseconds { 10 });
-   }
+         evt.Skip();
+   });
+   dialog.Bind(wxEVT_CLOSE_WINDOW, [this](wxCloseEvent& evt) {
+      evt.Skip();
+      mValidator.reset();
+      PluginManager::Get().Save();
+   });
+
+   dialog.CenterOnScreen();
+   ProcessNext();
+   dialog.ShowModal();
 }
 
 void PluginStartupRegistration::Stop()
 {
-   mValidator.reset();
-   PluginManager::Get().Save();
+   if(mScanDialog != nullptr)
+      mScanDialog->Close();
+}
+
+void PluginStartupRegistration::Skip()
+{
+   mValidator->SetDelegate(nullptr);
+   std::thread([validator = std::shared_ptr<AsyncPluginValidator>(std::move(mValidator))]{ }).detach();
+   mValidator = std::make_unique<AsyncPluginValidator>(*this);
+
+   if(!mValidProviderFound)
+   {
+      for(;mCurrentPluginProviderIndex < mPluginsToProcess[mCurrentPluginIndex].second.size(); ++mCurrentPluginProviderIndex)
+         OnPluginValidationFailed(
+            mPluginsToProcess[mCurrentPluginIndex].second[mCurrentPluginProviderIndex],
+            mPluginsToProcess[mCurrentPluginIndex].first);
+      mCurrentPluginProviderIndex = mPluginsToProcess[mCurrentPluginIndex].second.size() - 1;
+   }
+   OnValidationFinished();
```

Comment:
This function is the most difficult thing to understand in the entire pull request.  I don't think I do, yet.

It should be commented that the `OnValidationFinished` call supplies what would otherwise be an idle-time callback if the Skip button was not pressed.